### PR TITLE
RSDK-236 RSDK-248 - rpc: use single changestream per mongoDBWebRTCCallQueue

### PR DESCRIPTION
**Pull request by edaniels**
This improvement has been a long time coming. This makes using the MongoDB queue much more efficient while also adding in some nice soft-limits.

### CODE REVIEW

**Comment by edaniels** [comment]
@michaellee1019 I know this is a big PR, so take your time with it and don't feel rushed to get a review in. If there's anything you want to sit down about and discuss, let me know. Given this is your first time looking at this area of the code, I'd be happy to do an overview of what's going on here.

**Comment by edaniels** [comment]
FYI: I've tested this a ton locally and have gone through various disaster scenarios to see if this holds up and it appears to.

File: rpc/wrtc_call_queue_mongodb.go [hunk 1]
```diff
@@ -111,8 +277,532 @@ const (
...
+		projectStage,
+		unwindAggStage,
+		hostsMatch,
+		groupAggStage,
+	}
+	if forCaller {
+		pipeline = append(pipeline, queue.hostCallerQueueSizeMatchAggStage)
+	} else {
+		pipeline = append(pipeline, queue.hostAnswererQueueSizeMatchAggStage)
+	}
```
Comment (michaellee1019) [inline rpc/wrtc_call_queue_mongodb.go:791]: [opt] This is cool how we are using aggregations todo the logical comparison. But it took me a while to wrap my head around it. Might be more readable if we just do the comparison in go?
`if (forCaller && ret[0]['caller_size'] < maxHostCallers ){`
...

**Review by michaellee1019 — COMMENTED** [review]
Overall makes sense though still wrapping my head around some things. High level comments:

1. Consider an architecture diagram? Or a talk. Can be done later.
2. Consider running in staging only to let this bake for a while? It could help us review the operators collection against many robots to make sure everything is working as expected. I assume locally you have tested but with 1 caller and 1 answerer?

Some questions:
1. I assume that there will be a PR after this to implement on the App?
2. Is the plan to use a single operator per App host?
3. How can we monitor the worst case which is that all answerers are at capacity and no additional robots can connect?

**Comment by edaniels** [comment]
> Overall makes sense though still wrapping my head around some things. High level comments:
> 
>     1. Consider an architecture diagram? Or a talk. Can be done later.
Yes :). Not in this PR
> 
>     2. Consider running in staging only to let this bake for a while? It could help us review the operators collection against many robots to make sure everything is working as expected. I assume locally you have tested but with 1 caller and 1 answerer?
> 
I've tested with N:M caller:answerer but once this is in, we can put up a PR for app and deploy a staging env.
> 
> Some questions:
> 
>     1. I assume that there will be a PR after this to implement on the App?
> 
Yep, app needs to update to latest goutils and then it's in. App uses all the code rewritten here for signaling.
>     2. Is the plan to use a single operator per App host?
Yes
> 
>     3. How can we monitor the worst case which is that all answerers are at capacity and no additional robots can connect?
The metrics would allow us to check this. Keep in mind answering is a few seconds. You can have 100 active connections to a robot and still 2 answers processing new connections (think a switchboard)

File: rpc/wrtc_call_queue_mongodb.go [hunk 2]
```diff
@@ -4,42 +4,136 @@ import (
...
+	mongodbWebRTCCallQueueOperatorsCollName  = "operators"
+	mongodbWebRTCCallQueueRPCCallExpireName  = "rpc_call_expire"
+	mongodbWebRTCCallQueueOperatorExpireName = "operator_expire"
+)
+
+// this probably matches defaultMaxAnswerers on the signaling answerer.
+const maxHostAnswerersSize = 2
+
+// NewMongoDBWebRTCCallQueue returns a new MongoDB based call queue where calls are transferred
+// through the given client. The operator ID must be unique (e.g. a hostname, container ID, UUID, etc.).
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:116]: The operatorID can be a hostname. This function does not care. In app for example, we will use a cloud run instance ID. An operator is always 1 per host and multi purpose as well.

File: rpc/wrtc_call_queue_mongodb.go [hunk 3]
```diff
@@ -50,30 +144,98 @@ var (
...
+	}
+	if err := mongoutils.EnsureIndexes(ctx, operatorsColl, mongodbWebRTCCallQueueOperatorsIndexes...); err != nil {
 		return nil, err
 	}
+
+	if _, err := operatorsColl.InsertOne(ctx, bson.D{
+		{webrtcOperatorIDField, operatorID},
+		{webrtcOperatorExpireAtField, time.Now().Add(operatorHeartbeatWindow)},
+	}); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:184]: It's extremely unlikely and will cause startup failure so I don't think it needs to be logged.

File: rpc/wrtc_call_queue_mongodb.go [hunk 4]
```diff
@@ -111,8 +277,532 @@ const (
...
+
+func (queue *mongoDBWebRTCCallQueue) processClusterEventState(
+	newResumeToken bson.Raw,
+	newClusterTime primitive.Timestamp,
+) bool {
+	queue.csStateMu.Lock()
+	if !newClusterTime.IsZero() {
+		if queue.csLastEventClusterTime.T > newClusterTime.T ||
+			(queue.csLastEventClusterTime.T == newClusterTime.T &&
+				queue.csLastEventClusterTime.I >= newClusterTime.I) {
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:507]: It is not possible. Each "event" increments a logical clock in the database, which is this time. Transactions make things different but that's a non-issue here.

File: rpc/wrtc_call_queue_mongodb.go [hunk 5]
```diff
@@ -111,8 +277,532 @@ const (
...
+		queue.csManagerSeq.Add(1)
+		return true
+	}
+
+	if next.Error != nil {
+		queue.logger.Errorw("error getting next event in change stream", "error", next.Error)
+
+		if errors.Is(next.Error, mongoutils.ErrChangeStreamInvalidateEvent) {
+			queue.processClusterEventState(next.ResumeToken, primitive.Timestamp{})
+		}
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:539]: it's not very expected. It indicates something bad has happened

File: rpc/wrtc_call_queue_mongodb.go [hunk 6]
```diff
@@ -4,42 +4,136 @@ import (
...
+		Unit:        units.Dimensionless,
+		Labels: []statz.Label{
+			{Name: "operator_id", Description: "The queue operator ID."},
+		},
+	})
+)
+
 // A mongoDBWebRTCCallQueue is an MongoDB implementation of a call queue designed to be used for
 // multi-node, distributed deployments.
 type mongoDBWebRTCCallQueue struct {
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:69]: yeah, later. I personally don't think all fields are very useful to comment versus look at their usage

File: rpc/wrtc_call_queue_mongodb.go [hunk 7]
```diff
@@ -111,8 +277,532 @@ const (
...
+	// This message sending pass must be as fast possible and for that reason we use select defaults.
+	// In the event default cases happen, we have determined to either have hit some terminal case
+	// in the exchange or too many messages have been sent. Either way we should log and monitor these.
+	func() {
+		queue.csStateMu.RLock()
+		defer queue.csStateMu.RUnlock()
+
+		if next.Event.OperationType == mongoutils.ChangeEventOperationTypeInsert {
+			if _, ok := queue.csTrackingHosts[callResp.Host]; !ok {
+				// not interestsed in this
```
[the remark this answers is not in the document]
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:575]: I'll change it to "no one is currently subscribed to insert events for this host"

File: rpc/wrtc_call_queue_mongodb.go (hunk 1 above)
Comment (edaniels) [reply rpc/wrtc_call_queue_mongodb.go:791]: that puts more burden on us and increases the chance of fetching documents that do not need to be returned to us (network + compute costs), so I think it's better to keep in the DB.

File: rpc/wrtc_call_queue_mongodb.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaellee1019) [reply rpc/wrtc_call_queue_mongodb.go]: Ah I get it now. Thanks!

**Comment by michaellee1019** [comment]
> > Overall makes sense though still wrapping my head around some things. High level comments:
> > ```
> > 1. Consider an architecture diagram? Or a talk. Can be done later.
> > ```
> 
> Yes :). Not in this PR
> 
> > ```
> > 2. Consider running in staging only to let this bake for a while? It could help us review the operators collection against many robots to make sure everything is working as expected. I assume locally you have tested but with 1 caller and 1 answerer?
> > ```
> 
> I've tested with N:M caller:answerer but once this is in, we can put up a PR for app and deploy a staging env.
> 
> > Some questions:
> > ```
> > 1. I assume that there will be a PR after this to implement on the App?
> > ```
> 
> Yep, app needs to update to latest goutils and then it's in. App uses all the code rewritten here for signaling.
> 
> > ```
> > 2. Is the plan to use a single operator per App host?
> > ```
> 
> Yes
> 
> > ```
> > 3. How can we monitor the worst case which is that all answerers are at capacity and no additional robots can connect?
> > ```
> 
> The metrics would allow us to check this. Keep in mind answering is a few seconds. You can have 100 active connections to a robot and still 2 answerers processing new connections (think a distributed load balancer connected to N servers/listeners that keep running `accept`)

Makes sense! Thanks for explaining.

**Review by michaellee1019 — APPROVED** [review]

**Pull request merged**